internal/controller/numaflowcontroller/numaflowcontroller_controller.go

```diff
@@ -470,14 +441,35 @@ func (r *NumaflowControllerReconciler) sync(
 	controller *apiv1.NumaflowController,
 	namespace string,
 	numaLogger *logger.NumaLogger,
-	targetObjs []*unstructured.Unstructured,
 ) (gitopsSyncCommon.OperationPhase, error) {
 
-	reconciliationResult, diffResults, err := r.compareState(controller, namespace, targetObjs, numaLogger)
+	newVersion := controller.Spec.Version
+	newVersionTargetObjs, err := determineTargetObjects(controller, newVersion)
+	if err != nil {
+		return gitopsSyncCommon.OperationError, fmt.Errorf("unable to determine the target objects for the new version %s: %w", newVersion, err)
+	}
+
+	numaLogger.Debugf("found %d target objects associated with NumaflowController version %s", len(newVersionTargetObjs), newVersion)
+
+	reconciliationResult, diffResults, liveObjectsMap, err := r.compareState(controller, namespace, newVersionTargetObjs, numaLogger)
 	if err != nil {
 		return gitopsSyncCommon.OperationError, err
 	}
 
+	// Delete current resources if any of the specs differ
+	childResourcesNeedToBeDeleted := diffResults.Modified
```

Comment:
Actually, I'm kind of wondering if this puts us in an infinite loop? If we delete the resources below and requeue then do we come back through here, compare the desired objects to the deleted objects, decide they are different and then attempt to delete again?

Reply:
I do like the idea of each time checking if everything has been deleted. 

If my theory above is true about what's happening, one thought I have is that `childResourcesNeedToBeDeleted` could be based on whether there are in fact resources that need to be deleted. If `diffResults.Modified == true but all of the differences are deletions, then `childResourcesNeedToBeDeleted` could be set to false.

Reply:
As discussed over Slack, this solution did not work since liveObjectsMap is always empty (further investigation needed). Changed strategy by comparing the specs of the current version and the new version.